### PR TITLE
ci/test: improve CI jobs and tests

### LINKED ISSUE #1978: missing CI coverage
P.S.: CI needs quite the revamp, it's currently only covering Windows with no features and Linux with either all or no features;
@mattsse maybe just copy it over from Foundry?

_Originally posted by @DaniPopes in https://github.com/gakonst/ethers-rs/issues/1976#issuecomment-1365596397_

### LINKED ISSUE #1991: Move integration tests in 1 binary
Some integration tests (tests folder): https://github.com/gakonst/ethers-rs/tree/master/ethers-middleware/tests

use multiple binaries, this should be converted to this structure: https://github.com/foundry-rs/foundry/tree/master/anvil/tests/it

where each file is a module in: `tests/it/main.rs`

See also https://matklad.github.io/2022/07/04/unit-and-integration-tests.html

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Closes #1991
Closes #1978

## TODO

- [x] ethereum/feature tests
  - [x] Fix live/goerli tests
  - [x] foundry-rs/foundry#4431
  - [x] fix windows build
- [x] docs
  - [x] #2187
  - [x] #2203
- [x] examples
  - [x] #2207

Not planned for this PR (#2191):

- Feature checks: another PR because it requires massive feature changes. see: <https://github.com/gakonst/ethers-rs/actions/runs/4266867552/jobs/7427873846>
- WASM: same as feature checks. Doesn't work with --all-features because of tokio: <https://github.com/gakonst/ethers-rs/actions/runs/4267048448/jobs/7428242514>

## PR Checklist

-   [x] Added Tests
-   [x] Added Documentation
-   [ ] Updated the changelog
-   [ ] Breaking changes

### CODE REVIEW

File: ethers-core/src/types/transaction/eip2930.rs [hunk 1]
```diff
@@ -95,6 +95,7 @@ pub enum Eip2930RequestError {
 pub struct Eip2930TransactionRequest {
     #[serde(flatten)]
     pub tx: TransactionRequest,
+    #[serde(rename = "accessList")]
```
Comment (DaniPopes) [inline ethers-core/src/types/transaction/eip2930.rs:98]: i don't know if it was switching to anvil or what but typed_txs test failed sending the transaction with "unknown field" because this was not camelCase

File: ethers-middleware/Cargo.toml [hunk 2]
```diff
@@ -15,7 +15,9 @@ all-features = true
 rustdoc-args = ["--cfg", "docsrs"]
 
 [dependencies]
-ethers-contract = { version = "^1.0.0", path = "../ethers-contract", default-features = false }
+ethers-contract = { version = "^1.0.0", path = "../ethers-contract", default-features = false, features = [
+    "abigen",
+] }
```
Comment (DaniPopes) [inline ethers-middleware/Cargo.toml:20]: using abigen to generate DsProxyFactory (like multicall)

File: ethers-middleware/src/transformer/ds_proxy/mod.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-mod factory;
+pub mod factory;
```
Comment (DaniPopes) [inline ethers-middleware/src/transformer/ds_proxy/mod.rs:1]: make bindings public and also for int tests

File: ethers-middleware/tests/it/gas_escalator.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
+use ethers_core::{types::*, utils::Anvil};
+use ethers_middleware::{
+    gas_escalator::{Frequency, GasEscalatorMiddleware, GeometricGasPrice},
+    MiddlewareBuilder,
+};
+use ethers_providers::{Http, Middleware, Provider};
+use ethers_signers::{LocalWallet, Signer};
+
+#[tokio::test]
+#[ignore]
```
Comment (DaniPopes) [inline ethers-middleware/tests/it/gas_escalator.rs:10]: I don't know how gas escalator works, I tried porting the test to anvil but it was broken before too. I can't replicate the escalation

File: ethers-middleware/tests/it/main.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
+#![allow(clippy::extra_unused_type_parameters)]
```
Comment (DaniPopes) [inline ethers-middleware/tests/it/main.rs:1]: all ethers-middleware tests are now on Anvil, and the celo tests moved to tests/live/celo

File: ethers-middleware/tests/it/transformer.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+use ethers_signers::{LocalWallet, Signer};
+use rand::Rng;
+use std::sync::Arc;
+
+type HttpWallet = SignerMiddleware<Provider<Http>, LocalWallet>;
+
+abigen!(SimpleStorage, "../tests/testdata/SimpleStorage.json");
+
+#[tokio::test]
+#[ignore]
```
Comment (DaniPopes) [inline ethers-middleware/tests/it/transformer.rs:18]: This single test was not ignore before, but after using abigen the final assertions fail. I don't know how to fix it because I don't know what a DsProxy even is lol

File: ethers-providers/src/toolbox/pending_transaction.rs [hunk 7]
```diff
@@ -56,6 +56,7 @@ use wasm_timer::Delay;
 /// # }
 /// ```
 #[pin_project]
+#[must_use = "A PendingTransaction is a future and it must be polled or awaited to completion"]
```
Comment (DaniPopes) [inline ethers-providers/src/toolbox/pending_transaction.rs:59]: I think this is fitting since sometimes you might forget calling await a second time (`.send_transaction().await.unwrap()` -> `PendingTransaction` -> `.await.unwrap().unwrap()` -> `Receipt`)

File: examples/middleware/examples/nonce_manager.rs [hunk 8]
```diff
@@ -37,7 +37,7 @@ async fn main() -> Result<()> {
 
     assert_eq!(curr_nonce, 0);
 
-    nonce_manager.send_transaction(tx, None).await?;
+    nonce_manager.send_transaction(tx, None).await?.await?.unwrap();
```
Comment (DaniPopes) [inline examples/middleware/examples/nonce_manager.rs:40]: `PendingTransaction` must_use

File: tests/live/celo.rs [hunk 9]
```diff
@@ -0,0 +1,63 @@
+use crate::compile_contract;
```
Comment (DaniPopes) [inline tests/live/celo.rs:1]: the `#[cfg(feature = "celo")]` tests from `ethers-middleware::tests::signers`

File: tests/testdata/the_dao_abi.expr [hunk 10]
```diff
@@ -0,0 +1 @@
...ype\":\"uint256\"},{\"indexed\":false,\"name\":\"recipient\",\"type\":\"address\"},{\"indexed\":false,\"name\":\"amount\",\"type\":\"uint256\"},{\"indexed\":false,\"name\":\"newCurator\",\"type\":\"bool\"},{\"indexed\":false,\"name\":\"description\",\"type\":\"string\"}],\"name\":\"ProposalAdded\",\"type\":\"event\"},{\"anonymous\":false,\"inputs\":[{\"indexed\":true,\"name\":\"proposalID\",\"type\":\"uint256\"},{\"indexed\":false,\"name\":\"position\",\"type\":\"bool\"},{\"indexed\":true,\"name\":\"voter\",\"type\":\"address\"}],\"name\":\"Voted\",\"type\":\"event\"},{\"anonymous\":false,\"inputs\":[{\"indexed\":true,\"name\":\"proposalID\",\"type\":\"uint256\"},{\"indexed\":false,\"name\":\"result\",\"type\":\"bool\"},{\"indexed\":false,\"name\":\"quorum\",\"type\":\"uint256\"}],\"name\":\"ProposalTallied\",\"type\":\"event\"},{\"anonymous\":false,\"inputs\":[{\"indexed\":true,\"name\":\"_newCurator\",\"type\":\"address\"}],\"name\":\"NewCurator\",\"type\":\"event\"},{\"anonymous\":false,\"inputs\":[{\"indexed\":true,\"name\":\"_recipient\",\"type\":\"address\"},{\"indexed\":false,\"name\":\"_allowed\",\"type\":\"bool\"}],\"name\":\"AllowedRecipientChanged\",\"type\":\"event\"}]"
```
Comment (DaniPopes) [inline tests/testdata/the_dao_abi.expr:1]: extracted from `const DAO_ABI: &str = ...` in `ethers-etherscan::tests::contract`

File: ethers-etherscan/src/utils.rs [hunk 11]
```diff
@@ -3,8 +3,7 @@ use ethers_core::types::Address;
 use semver::Version;
 use serde::{Deserialize, Deserializer};
 
-static SOLC_BIN_LIST_URL: &str =
-    "https://raw.githubusercontent.com/ethereum/solc-bin/gh-pages/bin/list.txt";
+static SOLC_BIN_LIST_URL: &str = "https://binaries.soliditylang.org/bin/list.txt";
```
Comment (DaniPopes) [inline ethers-etherscan/src/utils.rs:6]: [ethereum/solc-bin](https://github.com/ethereum/solc-bin):
> The content is mirrored to https://binaries.soliditylang.org/ where it can be easily downloaded over HTTPS without any authentication, rate limiting or the need to use git.

File: ethers-providers/src/toolbox/pending_transaction.rs (hunk 7 above)
Comment (DaniPopes) [reply ethers-providers/src/toolbox/pending_transaction.rs]: moved to new branch

File: examples/middleware/examples/nonce_manager.rs (hunk 8 above)
Comment (DaniPopes) [reply examples/middleware/examples/nonce_manager.rs:40]: #2197

File: ethers-etherscan/src/utils.rs (hunk 11 above)
Comment (DaniPopes) [reply ethers-etherscan/src/utils.rs:6]: #2198

**Comment by DaniPopes** [comment]
@gakonst ready for review :pray:

Remaining failing actions:
- deny (licenses): will open an issue later / tomorrow
- windows: I have no clue what is wrong... It fails to compile a random example every time, which I cannot reproduce on my own machine
- examples: fixed by #2207

File: .github/scripts/install_test_binaries.sh [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    echo "Installed Geth:"
+    geth version
+    echo ""
+    echo "Installed Solc:"
+    solc --version
+}
+
+# Installs geth from https://geth.ethereum.org/downloads
+install_geth() {
+    case "$PLATFORM" in
```
Comment (prestwich) [inline .github/scripts/install_test_binaries.sh:38]: is it worth caching and then shortcutting if tool is already installed?

File: ethers-contract/ethers-contract-abigen/src/contract/events.rs [hunk 13]
```diff
@@ -230,19 +230,7 @@ pub(crate) fn event_struct_alias(event_name: &str) -> Ident {
 mod tests {
     use super::*;
     use crate::Abigen;
-    use ethers_core::abi::{EventParam, Hash, ParamType};
-    use proc_macro2::Literal;
-
-    /// Expands a 256-bit `Hash` into a literal representation that can be used with
```
Comment (prestwich) [inline ethers-contract/ethers-contract-abigen/src/contract/events.rs:236]: is this going somewhere? is dead code?

File: ethers-contract/ethers-contract-abigen/src/source/mod.rs [hunk 14]
```diff
@@ -141,17 +133,22 @@ mod tests {
...
-            "file://",
-            env!("CARGO_MANIFEST_DIR"),
-            "/../tests/solidity-contracts/console.json"
-        );
-        let exp = Source::Local(Path::new(rel).canonicalize().unwrap());
-        assert_eq!(Source::parse(rel).unwrap(), exp);
-        assert_eq!(Source::parse(abs).unwrap(), exp);
-        assert_eq!(Source::parse(abs_url).unwrap(), exp);
+        // Skip checking paths on Windows
+        #[cfg(not(windows))]
```
Comment (prestwich) [inline ethers-contract/ethers-contract-abigen/src/source/mod.rs:137]: should this be re-written to use [std::path](https://doc.rust-lang.org/std/path/index.html) instead of strings?

File: ethers-contract/tests/it/contract.rs [hunk 15]
```diff
@@ -1,951 +1,927 @@
-#![allow(unused)]
+use crate::common::*;
```
Comment (prestwich) [inline ethers-contract/tests/it/contract.rs:1]: this is a big diff. I'm guessing git is over-reporting a file rearrangement? can you summarize changes?

File: ethers-contract/tests/it/contract_call.rs [hunk 16]
```diff
@@ -6,9 +6,8 @@ use std::{
     sync::Arc,
 };
 
-#[tokio::test]
-async fn contract_call_into_future_is_send() {
-    abigen!(DsProxyFactory, "ethers-middleware/contracts/DsProxyFactory.json");
+fn _contract_call_into_future_is_send() {
```
Comment (prestwich) [inline ethers-contract/tests/it/contract_call.rs:9]: now unused?

File: ethers-core/src/utils/geth.rs [hunk 17]
```diff
@@ -351,13 +350,21 @@ impl Geth {
...
+        let bin_path = match self.program.as_ref() {
+            Some(bin) => bin.as_os_str(),
+            None => GETH.as_ref(),
+        }
+        .to_os_string();
+        let mut cmd = Command::new(&bin_path);
         // geth uses stderr for its logs
         cmd.stderr(Stdio::piped());
+
+        let mut unused_ports = unused_ports::<3>().into_iter();
```
Comment (prestwich) [inline ethers-core/src/utils/geth.rs:366]: this is good

File: .github/scripts/install_test_binaries.sh (hunk 12 above)
Comment (DaniPopes) [reply .github/scripts/install_test_binaries.sh:38]: hmm, would need to check version in case they differ, but we're only downloading binary so it's not that important imo

File: ethers-contract/tests/it/contract.rs (hunk 15 above)
Comment (DaniPopes) [reply ethers-contract/tests/it/contract.rs:1]: It's 99% indentation, contained only one module

File: ethers-contract/ethers-contract-abigen/src/contract/events.rs (hunk 13 above)
Comment (DaniPopes) [reply ethers-contract/ethers-contract-abigen/src/contract/events.rs:236]: I couldn't understand it; it's a local function in the test module, so it's not used anywhere else and had a test.. it doesn't really do anything

File: ethers-contract/tests/it/contract_call.rs (hunk 16 above)
Comment (DaniPopes) [reply ethers-contract/tests/it/contract_call.rs:9]: Just a compilation test

File: ethers-contract/ethers-contract-abigen/src/source/mod.rs (hunk 14 above)
Comment (DaniPopes) [reply ethers-contract/ethers-contract-abigen/src/source/mod.rs]: Realized that the diff was just the weird windows canonicalize thing, so dunce::canonicalize fixes that

File: ethers-middleware/tests/it/wallets.rs [hunk 18]
```diff
@@ -0,0 +1,62 @@
+static WALLETS: Lazy<TestWallets> = Lazy::new(|| {
```
Comment (gakonst) [inline ethers-middleware/tests/it/wallets.rs:1]: wait how is this working if it's importing nothing?

File: ethers-middleware/tests/it/transformer.rs (hunk 6 above)
Comment (gakonst) [reply ethers-middleware/tests/it/transformer.rs:18]: https://github.com/dapphub/ds-proxy we should understand why it's failing

File: .github/workflows/ci.yml [hunk 19]
```diff
@@ -1,288 +1,180 @@
...
-                  components: rustfmt, clippy
-                  override: true
-            - uses: Swatinem/rust-cache@v1
+            - uses: actions/checkout@v3
+            - uses: dtolnay/rust-toolchain@stable
+            - uses: Swatinem/rust-cache@v2
+            - name: live tests
+              run: cargo test -p ethers --test live --all-features
+
+    # TODO: [#2191](https://github.com/gakonst/ethers-rs/issues/2191)
```
Comment (gakonst) [inline .github/workflows/ci.yml:79]: sgtm - we can do as followup

**Review by gakonst — APPROVED** [review]
This all looks great to me. Would love if we can double click on why any of the new tests are failing, and if we ultimately think it's OK I'm down to track in a new issue and @DaniPopes can handle the next steps?

File: ethers-contract/ethers-contract-abigen/src/contract/events.rs [hunk 20]
```diff
@@ -406,20 +381,4 @@ mod tests {
             struct FooAliasedFilter(pub bool, pub ::ethers_core::types::Address);
         });
     }
-
-    #[test]
-    #[rustfmt::skip]
-    fn expand_hash_value() {
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply ethers-contract/ethers-contract-abigen/src/contract/events.rs:412]: yes https://github.com/gakonst/ethers-rs/pull/2189#discussion_r1119526731

File: .github/workflows/audit-on-push.yml [hunk 21]
```diff
@@ -1,14 +0,0 @@
-name: Security audit
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply .github/workflows/audit-on-push.yml:1]: see deps.yml file, same as reth

File: ethers-middleware/tests/it/wallets.rs (hunk 18 above)
Comment (DaniPopes) [reply ethers-middleware/tests/it/wallets.rs:1]: oh yeah i didn't want to delete it, but it's not used anywhere anymore since all middleware tests are on anvil now

File: ethers-middleware/tests/it/gas_escalator.rs (hunk 4 above)
Comment (DaniPopes) [reply ethers-middleware/tests/it/gas_escalator.rs:10]: maybe the assertions or the gas prices are wrong which I added, everything before that is the same just rewritten using anvil

File: ethers-middleware/tests/it/wallets.rs (hunk 18 above)
Comment (gakonst) [reply ethers-middleware/tests/it/wallets.rs:1]: ack - let's get rid of then

**Comment by gakonst** [comment]
OK. send it

**Comment by DaniPopes** [comment]
🫡
